docs/performance/optimization.rst

```diff
@@ -0,0 +1,544 @@
+large result sets in the range of several hundred thousand records, consider
+whether your application needs the whole result set at once. Use :ref:`cursors
+<server-side-cursor>` or ``LIMIT``/``OFFSET`` to fetch data in batches.
+
+See also `Fetching large result sets from CrateDB`_ for examples.
+
+.. _propagate-limit:
+
+*****************************************
+ Propagate LIMIT clauses when applicable
+*****************************************
+
+Similarly to the above, we may have for instance a ``LIMIT 10`` at the end of
+the query and to get there it may have been sufficient to only pull 10 records
+(or some other number of records) at an earlier stage from some given table, if
+that is the case duplicate or move (depending on specific query) the ``LIMIT``
+clause to the relevant place.
+
+In some cases, we may not know how many rows we need in the intermediate working
+sets but we may know for instance that for sure there will be 10 records on the
+last day of data, doing such filtering earlier on can be super helpful for the
+optimizer and may protect us from accidentally going through years of data.
+
+So for instance instead of:
+
+.. code:: sql
+
+   SELECT
+     factory_metadata.factory_name,
+     device_data.device_name,
+     device_data.reading_value
+   FROM device_data
+   INNER JOIN factory_metadata ON device_data.factory_id=factory_metadata.factory_id
+   WHERE reading_time BETWEEN '2024-01-01' AND '2025-01-01'
+   LIMIT 10;
+
+do:
+
+.. code:: sql
+
+   WITH filtered_device_data AS (
+      SELECT
+            device_data.factory_id,
+            device_data.device_name,
+            device_data.reading_value
+      FROM device_data
+      WHERE
+            reading_time BETWEEN '2024-12-01' AND '2025-01-01'
+            /* we are sure one month of data is sufficient to find 10 results and it may help with partition pruning */
+      LIMIT 10
+   )
+   SELECT
+     factory_metadata.factory_name,
+     filtered_device_data.device_name,
+     filtered_device_data.reading_value
+   FROM filtered_device_data
+   INNER JOIN factory_metadata ON filtered_device_data.factory_id=factory_metadata.factory_id;
+
+.. _only-sort-when-needed:
+
+****************************
+ Only sort data when needed
+****************************
+
+Indexing in CrateDB is optimized to support filtering and aggregations without
+requiring expensive defragmentation operations, but it is not optimized for
+sorting​.
+
+Maintaining a sorted index would slow down ingestion​ and Redshift and other
+analytical database systems like Cassandra make similar trade-offs​.
+
+This means that when an ``ORDER BY`` is requested the whole dataset comes in
+memory in a node and data is sorted there and then, hence it is important to not
+request ``ORDER BY`` operations when not actually needed, there is, of course,
+no problem sorting a few thousand rows in the final stage of a ``SELECT`` but we
+need to avoid requesting sort operations over millions of rows.
```

Comment:
```suggestion
This means that when an ``ORDER BY`` operation is requested, the whole dataset
needs to be loaded into the main memory on the relevant cluster node to be sorted.
That is why it is important to not request ``ORDER BY`` operations when not actually
needed, and most importantly, not on tables of large cardinalities without aggregating
records beforehand. On the other hand, of course it is no problem to sort a few
thousand rows in the final stage of a ``SELECT`` operation, but we need to avoid
requesting sort operations over millions of rows.
```